dll/win32/shell32/wine/shell32_main.c

```diff
@@ -489,31 +527,26 @@ DWORD_PTR WINAPI SHGetFileInfoW(LPCWSTR path,DWORD dwFileAttributes,
 
     if (flags & SHGFI_PIDL)
     {
-        pidl = ILClone((LPCITEMIDLIST)path);
-    }
-    else if (!(flags & SHGFI_USEFILEATTRIBUTES))
-    {
-        hr = SHILCreateFromPathW(szFullPath, &pidl, &dwAttributes);
+        pidl = (LPITEMIDLIST)path;
+        hr = pidl ? S_OK : E_FAIL;
     }
-
-    if ((flags & SHGFI_PIDL) || !(flags & SHGFI_USEFILEATTRIBUTES))
+    else
     {
-        /* get the parent shellfolder */
-        if (pidl)
+        if (flags & SHGFI_USEFILEATTRIBUTES)
         {
-            hr = SHBindToParent( pidl, &IID_IShellFolder, (LPVOID*)&psfParent,
-                                (LPCITEMIDLIST*)&pidlLast );
-            if (SUCCEEDED(hr))
-                pidlLast = ILClone(pidlLast);
-            else
-                hr = S_OK;
-            ILFree(pidl);
+            pidl = SHELL32_CreateSimpleIDListFromPath(szFullPath, dwFileAttributes);
+            hr = pidl ? S_OK : E_FAIL;
         }
         else
         {
-            ERR("pidl is null!\n");
-            return FALSE;
+            hr = SHILCreateFromPathW(szFullPath, &pidl, &dwAttributes);
         }
+        pidlFree = pidl;
+    }
+
+    if (SUCCEEDED(hr))
+    {
+        hr = SHBindToParent(pidl, &IID_IShellFolder, (void**)&psfParent, (LPCITEMIDLIST*)&pidlLast);
```

Comment:
```suggestion
        hr = SHBindToParent(pidl, IID_PPV_ARG(IShellFolder, &psfParent), (LPCITEMIDLIST*)&pidlLast);
```
and similar elsewhere